worlds/pokemon_emerald/locations.py

```diff
@@ -84,41 +153,68 @@ def create_location_label_to_id_map() -> Dict[str, int]:
     for region_data in data.regions.values():
         for location_name in region_data.locations:
             location_data = data.locations[location_name]
-            label_to_id_map[location_data.label] = offset_flag(location_data.flag)
+
+            if location_data.flag == 0:
+                label_to_id_map[location_data.label] = BASE_OFFSET + POKEDEX_OFFSET + int(location_data.name[15:])
+            else:
+                label_to_id_map[location_data.label] = offset_flag(location_data.flag)
 
     return label_to_id_map
 
 
-LOCATION_GROUPS = {
-    "Badges": {
-        "Rustboro Gym - Stone Badge",
-        "Dewford Gym - Knuckle Badge",
-        "Mauville Gym - Dynamo Badge",
-        "Lavaridge Gym - Heat Badge",
-        "Petalburg Gym - Balance Badge",
-        "Fortree Gym - Feather Badge",
-        "Mossdeep Gym - Mind Badge",
-        "Sootopolis Gym - Rain Badge",
-    },
-    "Gym TMs": {
-        "Rustboro Gym - TM39 from Roxanne",
-        "Dewford Gym - TM08 from Brawly",
-        "Mauville Gym - TM34 from Wattson",
-        "Lavaridge Gym - TM50 from Flannery",
-        "Petalburg Gym - TM42 from Norman",
-        "Fortree Gym - TM40 from Winona",
-        "Mossdeep Gym - TM04 from Tate and Liza",
-        "Sootopolis Gym - TM03 from Juan",
-    },
-    "Postgame Locations": {
-        "Littleroot Town - S.S. Ticket from Norman",
-        "SS Tidal - Hidden Item in Lower Deck Trash Can",
-        "SS Tidal - TM49 from Thief",
-        "Safari Zone NE - Item on Ledge",
-        "Safari Zone NE - Hidden Item North",
-        "Safari Zone NE - Hidden Item East",
-        "Safari Zone SE - Item in Grass",
-        "Safari Zone SE - Hidden Item in South Grass 1",
-        "Safari Zone SE - Hidden Item in South Grass 2",
-    }
-}
+def set_free_fly(world: "PokemonEmeraldWorld") -> None:
+    # Set our free fly location
+    # If not enabled, set it to Littleroot Town by default
+    fly_location_name = "EVENT_VISITED_LITTLEROOT_TOWN"
+    if world.options.free_fly_location:
+        fly_location_name = world.random.choice([
+            "EVENT_VISITED_SLATEPORT_CITY",
+            "EVENT_VISITED_MAUVILLE_CITY",
+            "EVENT_VISITED_VERDANTURF_TOWN",
+            "EVENT_VISITED_FALLARBOR_TOWN",
+            "EVENT_VISITED_LAVARIDGE_TOWN",
+            "EVENT_VISITED_FORTREE_CITY",
+            "EVENT_VISITED_LILYCOVE_CITY",
+            "EVENT_VISITED_MOSSDEEP_CITY",
+            "EVENT_VISITED_SOOTOPOLIS_CITY",
+            "EVENT_VISITED_EVER_GRANDE_CITY",
+        ])
+
+    world.free_fly_location_id = VISITED_EVENT_NAME_TO_ID[fly_location_name]
+
+    free_fly_location_location = world.multiworld.get_location("FREE_FLY_LOCATION", world.player)
+    free_fly_location_location.item = None
```

Comment:
is the item you're removing here a different event placed using `place_locked_item`? if so, i think it makes more sense to just do `free_fly_location_location.item = world.create_event(fly_location_name)`

Reply:
Yes, it is.

`place_locked_item` also gives `item.location = self` I guess. Of the two versions, I think I still prefer calling `place_locked_item`, but I truly have no strong opinion on it. I assume I'd want to also set the location on the new item. But if that still sounds more sensible to you, I'll change it.

Reply:
ah right good point. maybe `multiworld.push_item` then? it doesn't have the assert, and the location is already locked, which won't get changed doing this. bit of a hard call to make, since what you're currently doing is "more clear" on reading, but it's also a bit ugly. just a suggestion feel free to ignore 🤷

Reply:
Eh. I think I'll keep it the way it is. Doesn't look ugly enough to bother me.